samples/file-encryptor/enclave/openssl_src/keys.cpp

```diff
@@ -0,0 +1,438 @@
+    {
+        if (!(ret = EVP_EncryptInit_ex(
+                  ctx, EVP_aes_256_cbc(), NULL, encrypt_key, iv)))
+        {
+            TRACE_ENCLAVE("EVP_EncryptInit_ex failed with return code %d", ret);
+            goto exit;
+        }
+
+        EVP_CIPHER_CTX_set_padding(ctx, 0);
+        EVP_CIPHER_CTX_set_key_length(ctx, ENCRYPTION_KEY_SIZE_IN_BYTES);
+
+        if (!(ret = EVP_EncryptUpdate(
+                  ctx,
+                  output_data,
+                  (int*)&output_data_size,
+                  input_data,
+                  (int)input_data_size)))
+        {
+            TRACE_ENCLAVE("EVP_EncryptUpdate failed return code %d", ret);
+            goto exit;
+        }
+
+        if (!(ret = EVP_EncryptFinal_ex(
+                  ctx,
+                  output_data + output_data_size,
+                  &last_cipher_block_length)))
+        {
+            TRACE_ENCLAVE(
+                "EVP_EncryptFinal_ex failed with return code %d", ret);
+            goto exit;
+        }
+    }
+    else
+    {
+        if (!(ret = EVP_DecryptInit_ex(
+                  ctx, EVP_aes_256_cbc(), NULL, encrypt_key, iv)))
+        {
+            TRACE_ENCLAVE("EVP_DecryptInit_ex failed with return code %d", ret);
+            goto exit;
+        }
+
+        EVP_CIPHER_CTX_set_padding(ctx, 0);
+        EVP_CIPHER_CTX_set_key_length(ctx, ENCRYPTION_KEY_SIZE_IN_BYTES);
+
+        if (!(ret = EVP_DecryptUpdate(
+                  ctx,
+                  output_data,
+                  (int*)&output_data_size,
+                  input_data,
+                  (int)input_data_size)))
+        {
+            TRACE_ENCLAVE("EVP_DecryptUpdate failed with return code %d", ret);
+            goto exit;
+        }
+
+        if (!(ret = EVP_DecryptFinal_ex(
+                  ctx,
+                  output_data + output_data_size,
+                  &last_cipher_block_length)))
+        {
+            TRACE_ENCLAVE("EVP_DecryptFinal_ex failedwith return code %d", ret);
+            goto exit;
+        }
+    }
+
+    ret = 0;
+exit:
+    EVP_CIPHER_CTX_free(ctx);
+    return ret;
+}
+
+// For an encryption operation, the encryptor creates encryption metadata for
+// writing back to the encryption header, which includes the following fields:
+// digest: a hash value of the password
+// key: encrypted version of the encryption key
+//
+// Operations involves the following operations:
+//  1)derive a key from the password
+//  2)produce a encryption key
+//  3)generate a digest for the password
+//  4)encrypt the encryption key with a password key
+//
+int ecall_dispatcher::prepare_encryption_header(
+    encryption_header_t* header,
+    string password)
+{
+    int ret = 1;
+    unsigned char digest[HASH_VALUE_SIZE_IN_BYTES]; // sha256 digest of password
+    unsigned char
+        password_key[ENCRYPTION_KEY_SIZE_IN_BYTES]; // password derived key
+    unsigned char
+        encrypted_key[ENCRYPTION_KEY_SIZE_IN_BYTES]; // encrypted encryption key
```

Comment:
Why would the encryption key, the encrypted encryption key, and the password derived key all have to be the same length?  This at least warrants a comment explaining.

Reply:
added comments.